### PR TITLE
Update training logic to computes loss over logits

### PR DESCRIPTION
**Pull request by anwai98**
WIP!

### CODE REVIEW

**Comment by anwai98** [comment]
Hi @constantinpape,

This should be good for a first review from my side!

The intended idea here is simple: 1) compute loss over logit masks (to make this possible, I convert the labels to match the logits dimensions) and 2) work with iterative prompts for logits

Let me know how it looks!

File: micro_sam/training/sam_trainer.py [hunk 1]
```diff
@@ -240,11 +235,11 @@ def _compute_iterative_loss(self, batched_inputs, y_one_hot, num_subiter, multim
...
         return loss, mask_loss, iou_regression_loss, mean_model_iou
 
-    def _update_prompts(self, batched_inputs, y_one_hot, masks, logits_masks):
+    def _update_prompts(self, batched_inputs, y_one_hot, logits_masks):
         # here, we get the pair-per-batch of predicted and true elements (and also the "batched_inputs")
-        for x1, x2, _inp, logits in zip(masks, y_one_hot, batched_inputs, logits_masks):
+        for y_, _inp, logits in zip(y_one_hot, batched_inputs, logits_masks):
             # here, we get each object in the pairs and do the point choices per-object
-            net_coords, net_labels, _, _ = self.prompt_generator(x2, x1)
+            net_coords, net_labels, _, _ = self.prompt_generator(segmentation=y_, prediction=logits)
```
Comment (anwai98) [inline micro_sam/training/sam_trainer.py:242]: I observed one thing instantly in one of the training runs: small objects get smaller, hence it throws errors for not being able to find pixels to fetch points during iterative prompting. I will investigate this closely and see what can be done.

Otherwise, we might have to stick to full masks for iterative prompting!

File: micro_sam/training/joint_sam_trainer.py [hunk 2]
```diff
@@ -175,25 +172,22 @@ class JointSamLogger(TorchEmLogger):
...
-        self.log_dir = f"./logs/{trainer.name}" if save_root is None else\
-            os.path.join(save_root, "logs", trainer.name)
+        self.log_dir = f"./logs/{trainer.name}" if save_root is None else os.path.join(save_root, "logs", trainer.name)
         os.makedirs(self.log_dir, exist_ok=True)
 
         self.tb = SummaryWriter(self.log_dir)
         self.log_image_interval = trainer.log_image_interval
 
-    def add_image(self, x, y, samples, name, step):
+    def add_image(self, x, y, name, step):
```
Comment (constantinpape) [inline micro_sam/training/joint_sam_trainer.py:181]: Was this never used / not used anymore?

File: micro_sam/training/sam_trainer.py (hunk 1 above)
Comment (constantinpape) [reply micro_sam/training/sam_trainer.py:242]: But fundamentally these must still be there. Because it just does a naive upsampling of the low-res mask in SAM from what I remember, so it cannot add labels that where not there in the low-res mask?!
Maybe we just need to adapt some size threshold?

File: micro_sam/training/joint_sam_trainer.py (hunk 2 above)
Comment (anwai98) [reply micro_sam/training/joint_sam_trainer.py:181]: It was used previously, I removed it now because the `samples` are logits now. Would you recommend to keep those predictions by upsampling them?

File: micro_sam/training/sam_trainer.py (hunk 1 above)
Comment (anwai98) [reply micro_sam/training/sam_trainer.py:242]: Okay, even after a size sampler in place, the issue originates from the spot for finding positive point prompts in iterative prompting for a few objects. I will try to investigate what's the core reason behind it to get a better idea!

File: micro_sam/training/joint_sam_trainer.py (hunk 2 above)
Comment (constantinpape) [reply micro_sam/training/joint_sam_trainer.py:181]: What exactly were the samples? Examples for mask predictions? It would maybe be good to keep them, we can discuss later.

Comment (anwai98) [reply micro_sam/training/joint_sam_trainer.py:181]: Yes, samples were mask predictions!

File: micro_sam/training/sam_trainer.py (hunk 1 above)
Comment (anwai98) [reply micro_sam/training/sam_trainer.py:241]: This issue is more or less identified and probably is fixed with the latest commit (I will push a test training run anyways). For now, I'll resolve this discussion!

**Comment by anwai98** [comment]
Here's a detailed description on the PR (we see when we can come back to this in future)

Core Idea: Computing loss over predicted logits masks and downsampled ground-truth masks.

The current implementation works, however it does not bring us significant memory advantages (below mentioned are quick try-outs on LIVECell):

```
# # for default settings
# loss over logits: 47.89 GB
# loss over masks: 49.42 GB

# # for resource-efficient setting (i.e. n_objects=5 for 'vit_b')
# loss over logits: 24.01 GB
# loss over masks: 24.38 GB
```
